### PR TITLE
MODINV-820: Implement sharing of Instance with source=FOLIO

### PR DESCRIPTION
**Pull request by Aliaksandr-Fedasiuk**
## Purpose
Implement sharing of Instance with source=FOLIO

##Approach

- Implement consumer for CONSORTIUM_INSTANCE_SHARING_INIT event
- Verify absence of shared Instance - make GET request by Instance UUID on target (consortium) tenant, if exists - (publish error event?), if not - proceed
- If source FOLIO - make GET request by Instance UUID on source tenant and POST on target tenant with source=FOLIO
- Convert local Instance in Shadow - make PUT request to update source to CONSORTIUM-FOLIO, set HRID (changing logic of PUT endpoint is our of scope of this task)
- Publish CONSORTIUM_INSTANCE_SHARING_COMPLETE event if successful, ERROR (CONSORTIUM_INSTANCE_SHARING_COMPLETE) if error

## Learning
[MODINV-820](https://issues.folio.org/browse/MODINV-820)

### CODE REVIEW

File: src/main/java/org/folio/inventory/ConsortiumInstanceSharingConsumerVerticle.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    KafkaConfig kafkaConfig = KafkaConfig.builder()
+      .envId(config.getString(KAFKA_ENV))
+      .kafkaHost(config.getString(KAFKA_HOST))
+      .kafkaPort(config.getString(KAFKA_PORT))
+      .okapiUrl(config.getString(OKAPI_URL))
+      .replicationFactor(Integer.parseInt(config.getString(KAFKA_REPLICATION_FACTOR)))
+      .maxRequestSize(Integer.parseInt(config.getString(KAFKA_MAX_REQUEST_SIZE)))
+      .build();
+    LOGGER.info(format("kafkaConfig: %s", kafkaConfig));
+    EventManager.registerKafkaEventPublisher(kafkaConfig, vertx, maxDistributionNumber);
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/ConsortiumInstanceSharingConsumerVerticle.java:51]: With my limited knowledge of the app. It seems the line would be called somewhere else already. There is nothing unique about the kafkaConfig class here.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+import static org.apache.commons.lang.StringUtils.isNotBlank;
+import static org.folio.inventory.consortium.model.ConsortiumEnumStatus.COMPLETE;
+import static org.folio.inventory.consortium.model.ConsortiumEnumStatus.ERROR;
+import static org.folio.inventory.dataimport.util.DataImportConstants.UNIQUE_ID_ERROR_MESSAGE;
+
+public class ConsortiumInstanceSharingHandler implements AsyncRecordHandler<String, String> {
+
+  private static final Logger LOGGER = LogManager.getLogger(DataImportKafkaHandler.class);
+
+  private static final String OKAPI_TOKEN_HEADER = "x-okapi-token";
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:50]: I believe that are some constants in RMB that already store these values. It would be good to reuse them.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        kafkaHeaders.get(OKAPI_TOKEN_HEADER), kafkaHeaders.get(OKAPI_URL_HEADER));
+      InstanceCollection targetInstanceCollection = storage.getInstanceCollection(targetTenantContext);
+
+      Context sourceTenantContext = EventHandlingUtil.constructContext(sharingInstance.getSourceTenantId(),
+        kafkaHeaders.get(OKAPI_TOKEN_HEADER), kafkaHeaders.get(OKAPI_URL_HEADER));
+      InstanceCollection sourceInstanceCollection = storage.getInstanceCollection(sourceTenantContext);
+
+      LOGGER.info("handle :: checking is instance {} exists on tenant {}", instanceId, sharingInstance.getTargetTenantId());
+      getInstanceById(instanceId, sharingInstance.getTargetTenantId(), targetInstanceCollection)
+        .onFailure(failure -> {
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:93]: The chain of handlers here is too long(all the way to the end of this method). Please break it out into private methods to improve readability.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+          if (failure.getClass().equals(NotFoundException.class)) {
+            LOGGER.info("handle :: instance {} not found on target tenant: {}", instanceId, sharingInstance.getTargetTenantId());
+            getInstanceById(instanceId, sharingInstance.getSourceTenantId(), sourceInstanceCollection)
+              .onSuccess(srcInstance -> {
+                if (srcInstance == null) {
+                  String errorMessage = format("instance %s not found on source tenant: %s", instanceId, sharingInstance.getSourceTenantId());
+                  sendCompleteEventToKafka(tenantId, sharingInstance, kafkaConfig, record.headers());
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:96]: I can't tell if the complete event will have an error within...

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+                  promise.fail(errorMessage);
+                } else {
+                  LOGGER.info("handle :: Publishing instance {} with source {} from {} tenant to {} tenant", instanceId, srcInstance.getSource(), sharingInstance.getSourceTenantId(), sharingInstance.getTargetTenantId());
+                  if ("FOLIO".equals(srcInstance.getSource())) {
+                    publishInstanceToTenant(srcInstance, targetInstanceCollection)
+                      .onSuccess(publishedInstance -> {
+                        LOGGER.info("handle :: Updating source to 'CONSORTIUM-FOLIO' for Instance {}", instanceId);
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:103]: Every reference to CONSORTIUM-FOLIO should come from an enum or constant rather than a raw string.

Comment (okolawole-ebsco) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:103]: in the info message, indicate the the UUID provided is the instance id. Example `InstanceId=<UUID>`

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+                        JsonObject jsonInstanceToPublish = srcInstance.getJsonForStorage();
+                        jsonInstanceToPublish.put("source", "CONSORTIUM-FOLIO");
+                        updateInstanceInStorage(Instance.fromJson(jsonInstanceToPublish), sourceInstanceCollection)
+                          .onSuccess(updatesSourceInstance -> {
+                            LOGGER.info("handle :: Source 'CONSORTIUM-FOLIO' updated for Instance {}", instanceId);
+                            sendCompleteEventToKafka(tenantId, sharingInstance, kafkaConfig, record.headers());
+                            promise.complete();
+                          }).onFailure(error -> {
+                            String errorMessage = format("Error update Instance by id %s on the source tenant %s. Error: %s", instanceId, sharingInstance.getTargetTenantId(), error.getCause());
+                            sendErrorResponseAndPrintLogMessage(tenantId, errorMessage, sharingInstance, record.headers());
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:113]: Where is the error response sent to?

Comment (okolawole-ebsco) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:113]: Re-reading the code, it seems this is the negative/failure version of sendCompleteEventToKafka. I think sendCompleteEventToKafka should be setup such that it can send positive or negative.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:1]: I would prefer if log messages use "InstanceId=<UUID>" instead of "Instance by id <UUID>". This is easier searching of logs further down the road.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+                promise.fail(errorMessage);
+              });
+          } else {
+            String errorMessage = format("Error checking Instance by id %s on target tenant %s. Error: %s", instanceId, sharingInstance.getTargetTenantId(), failure.getMessage());
+            sendErrorResponseAndPrintLogMessage(tenantId, errorMessage, sharingInstance, record.headers());
+            promise.fail(errorMessage);
+          }
+        })
+        .onSuccess(instanceOnTargetTenant -> {
+          String errorMessage = format("Instance %s is present on target tenant: %s", instanceId, sharingInstance.getTargetTenantId());
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:147]: if the instance is already present, I think we should just send a complete message without error.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+  }
+
+  private void sendErrorResponseAndPrintLogMessage(String tenantId, String errorMessage, SharingInstance sharingInstance, List<KafkaHeader> kafkaHeaders) {
+    LOGGER.error("handle:: {}", errorMessage);
+    sendEventToKafka(tenantId, sharingInstance, ERROR, errorMessage, kafkaConfig, kafkaHeaders);
+  }
+
+  private Future<Instance> publishInstanceToTenant(Instance instance, InstanceCollection instanceCollection) {
+    JsonObject jsonInstance = instance.getJsonForStorage();
+    jsonInstance.remove("hrid");
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:165]: is `"hrid"` stored in some constant somewhere in mod-inventory?

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+      return Future.failedFuture(ex);
+    }
+  }
+
+  private void sendErrorResponseAndPrintLogMessage(String tenantId, String errorMessage, SharingInstance sharingInstance, List<KafkaHeader> kafkaHeaders) {
+    LOGGER.error("handle:: {}", errorMessage);
+    sendEventToKafka(tenantId, sharingInstance, ERROR, errorMessage, kafkaConfig, kafkaHeaders);
+  }
+
+  private Future<Instance> publishInstanceToTenant(Instance instance, InstanceCollection instanceCollection) {
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:163]: Based on the name of this method i was expecting tenant id to be one of the parameters. But this is more like adding an instance to the instance collection. I think the name of the method should reflect that and javadocs to explain further if needed.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    JsonObject jsonInstance = instance.getJsonForStorage();
+    jsonInstance.remove("hrid");
+    return addInstance(Instance.fromJson(jsonInstance), instanceCollection);
+  }
+
+  private Future<Instance> getInstanceById(String instanceId, String tenantId, InstanceCollection instanceCollection) {
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:169]: Thinking the logic within this method should already be included in `instanceCollection.findById` and there would be no need to declare this method.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+      },
+      failure -> {
+        LOGGER.error(format("getInstanceById :: Error retrieving Instance by id %s on tenant %s - %s, status code %s",
+          instanceId, tenantId, failure.getReason(), failure.getStatusCode()));
+        promise.fail(failure.getReason());
+      });
+    return promise.future();
+  }
+
+  private Future<Instance> addInstance(Instance instance, InstanceCollection instanceCollection) {
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:189]: change the method name to "addInstanceInStorage" since you already have "updateInstanceInStorage".

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+    sendEventToKafka(tenantId, sharingInstance, COMPLETE, null, kafkaConfig, kafkaHeaders);
+  }
+
+  private void sendEventToKafka(String tenantId, SharingInstance sharingInstance, ConsortiumEnumStatus status, String message,
+                                KafkaConfig kafkaConfig, List<KafkaHeader> kafkaHeaders) {
+
+    ConsortiumEvenType evenType = ConsortiumEvenType.CONSORTIUM_INSTANCE_SHARING_COMPLETE;
+
+    try {
+      LOGGER.info("sendEventToKafka :: tenantId: {}, instanceId: {}, status: {}{}",
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:230]: I think there should be a space between the last two {}

File: src/main/java/org/folio/inventory/consortium/model/ConsortiumEnumStatus.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
+package org.folio.inventory.consortium.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+public enum ConsortiumEnumStatus {
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/model/ConsortiumEnumStatus.java:6]: I think this should be called "SharingStatus" or similar.

Also there might be overlap on this enum with another PR from @RomanChernetskyi

File: src/main/java/org/folio/inventory/consortium/model/ConsortiumEvenType.java [hunk 14]
```diff
@@ -0,0 +1,24 @@
+package org.folio.inventory.consortium.model;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+
+public enum ConsortiumEvenType {
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/inventory/consortium/model/ConsortiumEvenType.java:5]: The name should reflect "instance" and "sharing". Unless you want to have an enum for all ECS events, then there is a typo - ConsortiumEventTypes

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 2 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:50]: I couldn't find these constants

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 3 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:89]: Thanks, It will be fixed in scope of [MODINV-870](https://issues.folio.org/browse/MODINV-870 )

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+            LOGGER.info("handle :: instance {} not found on target tenant: {}", instanceId, sharingInstance.getTargetTenantId());
+            getInstanceById(instanceId, sharingInstance.getSourceTenantId(), sourceInstanceCollection)
+              .onSuccess(srcInstance -> {
+                if (srcInstance == null) {
+                  String errorMessage = format("instance %s not found on source tenant: %s", instanceId, sharingInstance.getSourceTenantId());
+                  sendCompleteEventToKafka(tenantId, sharingInstance, kafkaConfig, record.headers());
+                  promise.fail(errorMessage);
+                } else {
+                  LOGGER.info("handle :: Publishing instance {} with source {} from {} tenant to {} tenant", instanceId, srcInstance.getSource(), sharingInstance.getSourceTenantId(), sharingInstance.getTargetTenantId());
+                  if ("FOLIO".equals(srcInstance.getSource())) {
```
[the remark this answers is not in the document]
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:100]: added enum org.folio.inventory.domain.instances.InstanceSource

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 4 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:96]: Sure. This message should inform the user that the instance in the source does not exist.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 5 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:103]: Completed.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 4 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:96]: Fixed. Copy-Past error. Thanks.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:1]: Fixed.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 7 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:147]: I sent a completed message with a warning description.

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 8 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:165]: Fixed

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 9 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:163]: Fixed. Renamed to publishInstanceToTenantSpecificCollection

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 11 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:189]: Thanks, It will be fixed in scope of [MODINV-870](https://issues.folio.org/browse/MODINV-870)

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 12 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:230]: Fixed.

File: src/main/java/org/folio/inventory/consortium/model/ConsortiumEnumStatus.java (hunk 13 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/model/ConsortiumEnumStatus.java:6]: Completed.

File: src/main/java/org/folio/inventory/consortium/model/ConsortiumEvenType.java (hunk 14 above)
Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/model/ConsortiumEvenType.java:5]: Fixed

File: src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java (hunk 2 above)
Comment (RuslanLavrov) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:50]: Olamide Aliaks, mod-inventory doesn't use RMB that's why these constants might not be present here

Comment (okolawole-ebsco) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:50]: I am able to reference it in mod-inventory code, so I assume it is pulled in by some transitive dependency.

Comment (Aliaksandr-Fedasiuk) [reply src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java:50]: image
